### ISSUE TITLE
Distributed wrapper for loss functions

### ISSUE THREAD

**Issue by KevinMusgrave** [KevinMusgrave/pytorch-metric-learning#180]
Try to make a DistributedLoss wrapper that makes loss functions work across multiple gpus, i.e. DistributedDataParallel.

@JohnGiorgi Maybe you have some thoughts. Is [this function in your library](https://github.com/JohnGiorgi/DeCLUTR/blob/1dda901fbfcfd9de5767daf06cbd5e8d1e81a86c/declutr/common/model_utils.py#L28) sufficient to make loss functions work with DistributedDataParallel?

**Comment by JohnGiorgi** [comment]
@KevinMusgrave Yes. I tested it it [a while ago](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/10), making sure that `num_neg_pairs` was as expected when training with `DistributedDataParallel` (as per [your suggestion](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/10#issuecomment-589971159)) (although more tests wouldn't hurt). I've also used this code extensively in my project without issues.

It assumes you are passing two tensors (`anchors` and `positives`), not sure if that will need to be modified to work with the loss functions of this library.

**Comment by KevinMusgrave** [comment]
K thanks I'll try it out. My guess is that I just need to replace the ```anchors``` and ```positives``` with ```embeddings```.

**Comment by AlexSchuy** [comment]
Hi @KevinMusgrave, I checked the documentation regarding losses, but it doesn't mention distributed training. Does this issue mean that currently, pytorch-metric-learning supports single gpu training only?

**Comment by KevinMusgrave** [comment]
@AlexSchuy You can wrap your model in DataParallel, meaning each batch will get split across multiple gpus. Everything in the library will work fine in this case. (Just don't wrap the losses, miners etc in DataParallel.)

What won't work properly is using DistributedDataParallel, because this will create separate processes. Each process operates on a part of the batch, and the loss in each process won't be aware of the elements in the other processes. So the loss won't be able to correctly form pairs/triplets.
